cpp/src/groupby/groupby.cu

```diff
@@ -108,6 +114,89 @@ void verify_valid_requests(std::vector<aggregation_request> const& requests) {
+ *   The total window size = `preceding_window + following_window + 1`. Element `i` uses elements
+ *   `[i-preceding_window, i+following_window]` to do the window computation, provided that they
+ *   fall within the confines of their corresponding groups, as indicated by `group_offsets`.
+ * - instead of storing NA/NaN for output rows that do not meet the minimum number of observations
+ *   this function updates the valid bitmask of the column to indicate which elements are valid.
+ * 
+ * The returned column for `op == COUNT` always has `INT32` type. All other operators return a 
+ * column of the same type as the input. Therefore it is suggested to convert integer column types
+ * (especially low-precision integers) to `FLOAT32` or `FLOAT64` before doing a rolling `MEAN`.
+ *
+ * @param[in] input_col The input column
+ * @param[in] group_offsets A column of indexes, indicating partition/grouping boundaries.
+ * @param[in] preceding_window The static rolling window size in the backward direction.
+ * @param[in] following_window The static rolling window size in the forward direction.
+ * @param[in] min_periods Minimum number of observations in window required to have a value,
+ *                        otherwise element `i` is null.
+ * @param[in] op The rolling window aggregation type (SUM, MAX, MIN, etc.)
+ *
+ * @returns   A nullable output column containing the rolling window results
+ **/
+
+std::unique_ptr<column> rolling_window(column_view const& input,
+                                       rmm::device_vector<cudf::size_type> const& group_offsets,
+                                       size_type preceding_window,
+                                       size_type following_window,
+                                       size_type min_periods,
+                                       std::unique_ptr<aggregation> const& aggr,
+                                       rmm::mr::device_memory_resource* mr)
+{
+  if (group_offsets.empty()) {
+    // Empty group_offsets list. Treat `input` as a single group. i.e. Ignore grouping.
+    return rolling_window(input, preceding_window, following_window, min_periods, aggr, mr);
+  }
+
+  // `group_offsets` are interpreted in adjacent pairs, each pair representing the offsets
+  // of the first, and one past the last elements in a group.
+  //
+  // If `group_offsets` is not empty, it must contain at least two offsets:
+  //   a. 0, indicating the first element in `input`
+  //   b. input.size(), indicating one past the last element in `input`.
+  //
+  // Thus, for an input of 1000 rows,
+  //   0. [] indicates a single group, spanning the entire column.
+  //   1  [10] is invalid.
+  //   2. [0, 1000] indicates a single group, spanning the entire column (thus, equivalent to no groups.)
+  //   3. [0, 500, 1000] indicates two equal-sized groups: [0,500), and [500,1000).
+
+  CUDF_EXPECTS(group_offsets.size() >= 2 && group_offsets[0] == 0 
```

Comment:
`group_offsets[0] == 0`
`group_offsets[group_offsets.size()-1]`

This is doing device to host memory copies. This is probably not what you want to be doing.

Reply:
Please pardon my lack of CUDA expertise. What's the best way to assert that the ends are capped with `0, max_offset`?

Reply:
Seeing as though this function is for internal use only, I think we can guarantee that the `group_offset` values will be within bounds and not require any bounds checking.

Reply:
I'd take it a step further and make the `group_offsets` be an iterator instead of a `device_vector`.

Reply:
Alright. Perhaps I could put that check in `#ifndef NDEBUG`?